### PR TITLE
Reversible cursors

### PR DESCRIPTION
**Pull request by ryzhyk**
We generalize cursor traits to iterate and search in both directions.

This is a step toward introducing support for SQL operators that manipulate
ordered groups of rows, including `ORDER BY ASC/DESC`, `RANK`,
`ROW_NUMBER`, `top-k`, etc.  These operators require the ability to
enumerate rows in both ascending and descending orders.  Our cursors
did not support that as they only allowed iterating and seeking forward.

This commit extends cursor traits with methods to iterate and seek
backward.  A cursor is always created pointing to the first record.  In
order to iterate backward, the client must first move the cursor to the
end of the range using the new `fast_forward()` method.

For batch and trace cursors we allow reverse iteration by both keys and
values.

So far the new infrastructure is only used to implement the `Max`
aggregate more efficiently.  Upcoming commits will make a more extensive
use of it.

### CODE REVIEW

File: demo/demo_notebooks/fraud_detection.ipynb [hunk 1]
```diff
@@ -30,7 +30,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 9,
+   "execution_count": 1,
```
Comment (mbudiu-vmw) [inline demo/demo_notebooks/fraud_detection.ipynb:33]: I am not sure what these changes have to do with the current PR.

File: crates/dbsp/src/trace/layers/advance.rs [hunk 2]
```diff
@@ -71,6 +71,85 @@ where
...
+
+    let last_index = slice.len() - 1;
+
+    // Exponential search if the answer isn't within `SMALL_LIMIT`.
+    if slice.len() > SMALL_LIMIT && function(&slice[last_index - SMALL_LIMIT]) {
+        // Skip `slice[..SMALL_LIMIT]` outright, the above condition established
+        // that nothing within it satisfies the predicate
+        let mut index = SMALL_LIMIT + 1;
+
+        // FIXME: This is kinda weird
```
Comment (mbudiu-vmw) [inline crates/dbsp/src/trace/layers/advance.rs:112]: what is weird?

File: crates/dbsp/src/trace/layers/column_layer/cursor.rs [hunk 3]
```diff
@@ -43,15 +43,33 @@ where
...
+            self.pos -= retreat(
+                &self.storage.keys[self.bounds.0..=self.pos as usize],
+                predicate,
+            ) as isize;
+        }
     }
 
     pub fn current_key(&self) -> &K {
-        &self.storage.keys[self.pos]
+        &self.storage.keys[self.pos as usize]
```
Comment (mbudiu-vmw) [inline crates/dbsp/src/trace/layers/column_layer/cursor.rs:68]: perhaps this is paranoia, but could this be checked for overflow?

Comment (mbudiu-vmw) [reply crates/dbsp/src/trace/layers/column_layer/cursor.rs:68]: I am guessing that pos is always positive, but there is no subtraction for usize?

File: crates/dbsp/src/trace/persistent/cursor.rs [hunk 4]
```diff
@@ -100,48 +115,58 @@ where
...
+            && self.val_idx >= 0
     }
 
     fn key(&self) -> &B::Key {
         self.cur_key.as_ref().unwrap()
     }
 
     fn val(&self) -> &B::Val {
-        &self.cur_vals.as_ref().unwrap()[self.val_idx].0
+        &self.cur_vals.as_ref().unwrap()[self.val_idx as usize].0
```
Comment (mbudiu-vmw) [inline crates/dbsp/src/trace/persistent/cursor.rs:161]: why isn't val_idx negative?

**Review by mbudiu-vmw — APPROVED** [review]
This stuff isn't easy to review, especially as changes.

File: crates/dbsp/src/operator/aggregate/mod.rs [hunk 5]
```diff
@@ -105,17 +105,17 @@ pub trait Aggregator<K, T, R>: Clone + 'static {
     ///
     /// * The method must return `None` if the total weight of each key is zero.
     ///   It must return `Some` otherwise.
-    fn aggregate<'s, C>(&self, cursor: &mut C) -> Option<Self::Accumulator>
+    fn aggregate<C>(&self, cursor: &mut C) -> Option<Self::Accumulator>
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/dbsp/src/operator/aggregate/mod.rs:108]: It's not.  I just wanted to make things a bit simpler before making them more complex :)

File: demo/demo_notebooks/fraud_detection.ipynb (hunk 1 above)
Comment (ryzhyk) [reply demo/demo_notebooks/fraud_detection.ipynb:33]: It's an accident, I'll remove this file from the PR.

File: crates/dbsp/src/trace/layers/column_layer/cursor.rs (hunk 3 above)
Comment (ryzhyk) [reply crates/dbsp/src/trace/layers/column_layer/cursor.rs:68]: I should add some comments about it. `pos` is `isize`, so we can represent an invalid cursor that moved beyond the left end of the range as `-1`. `current_key` is not allowed to be called on an invalid cursor. I will add debug checks for this.

File: crates/dbsp/src/trace/persistent/cursor.rs (hunk 4 above)
Comment (ryzhyk) [reply crates/dbsp/src/trace/persistent/cursor.rs:161]: Same here.

File: crates/dbsp/src/operator/time_series/partitioned.rs [hunk 6]
```diff
@@ -136,9 +150,33 @@ where
...
+    fn fast_forward_keys(&mut self) {
+        self.cursor.fast_forward_vals();
+        self.key = self.cursor.val().0.clone();
+    }
+
     fn rewind_vals(&mut self) {
         unimplemented!()
     }
+
+    fn step_val_reverse(&mut self) {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/dbsp/src/operator/time_series/partitioned.rs:162]: These methods are never used at the moment, but can be implemented if necessary.

File: crates/dbsp/src/trace/test_batch.rs [hunk 7]
```diff
@@ -438,9 +609,71 @@ where
...
+
+        while self.index + 1 < self.data.len() && self.data[self.index + 1].0 .0 == current_key {
+            self.index += 1;
+        }
+
+        let current_val = self.data[self.index].0 .1.clone();
+
+        while self.index > 0
+            && self.data[self.index - 1].0 .0 == current_key
+            && self.data[self.index - 1].0 .1 == current_val
```
Comment (Kixiron) [inline crates/dbsp/src/trace/test_batch.rs:670]:
```suggestion
            && self.data[self.index - 1].0 == (current_key, current_val)
```

File: crates/dbsp/src/trace/test_batch.rs [hunk 8]
```diff
@@ -438,9 +609,71 @@ where
...
+                self.val_valid = false;
+                return;
+            }
+        }
+
+        let current_val = self.data[self.index].0 .1.clone();
+
+        while self.index > 0
+            && self.data[self.index - 1].0 .0 == current_key
+            && self.data[self.index - 1].0 .1 == current_val
```
Comment (Kixiron) [inline crates/dbsp/src/trace/test_batch.rs:637]:
```suggestion
            && self.data[self.index - 1].0 == (current_key, current_val)
```

File: crates/dbsp/src/trace/test_batch.rs [hunk 9]
```diff
@@ -438,9 +609,71 @@ where
...
+        todo!()
+    }
+
     fn rewind_vals(&mut self) {
         todo!()
     }
+
+    fn step_val_reverse(&mut self) {
+        let current_key = self.data[self.index].0 .0.clone();
+        let current_val = self.data[self.index].0 .1.clone();
```
Comment (Kixiron) [inline crates/dbsp/src/trace/test_batch.rs:622]:
```suggestion
        let (current_key, current_val) = self.data[self.index].0.clone();
```

**Review by Kixiron — APPROVED** [review]
We probably want a reversible interface for `Consumer` as well

File: crates/dbsp/src/trace/test_batch.rs (hunk 7 above)
Comment (ryzhyk) [reply crates/dbsp/src/trace/test_batch.rs:670]: This won't do, there's also the time field in that tuple.

**Pull request merged**